KT-52951

mpetuska opened the pull request:
Configurable js test browser target

Closes [KT-52951](https://youtrack.jetbrains.com/issue/KT-52951)

Usage:
```kotlin
// build.gradle.kts
plugins{
  kotlin("multiplatform")
}

kotlin {
  js("MyTargetName") {
    browser {
      testTask {
        useKarma {}
      }
    }
  }
}
```
```properties
# gradle.properties
kotlin.target.MyTargetName.js.test.browsers=firefox-headless,chrome-headless,ie
```
```properties
# local.properties
kotlin.target.MyTargetName.js.test.browsers=firefox
```

mpetuska commented:
Added multiple browser support via comma separated list

mpetuska commented:
Ready for review

mpetuska commented:
Btw, feel free to suggest a better property key. Maybe specifying target type (js) after target name is redundant?

mpetuska commented:
A gentle reminder to have a look at this before I go ahead solving new merge conflicts :)

ilgonmic commented:
Possibly it is more logical to make similar with DSL structure
Something like
```
kotlin.js.<target-name>?.browser.karma.browsers
```
and for default/main 
```
kotlin.js.browser.karma.browsers
```

And another mind, that in KGP we don’t have mechanism to undo particular browser.
So if in gradle.properties there is Chrome, for now it can’t be canecelled.
We don’t have something like `notUseChrome()`. Is it ok for your use case?

But possibly it can be undo with `gradle.properties` on subproject level
Something like
Root `gradle.properties` (using such browsers for all projects)
```
kotlin.target.MyTargetName.js.test.browsers=firefox-headless,chrome-headless,ie
```

Subproject `gradle.properties` (override browsers for particular subproject). But should we have possibility to set `none` browser?
```
kotlin.target.MyTargetName.js.test.browsers=chrome-headless
```

mpetuska commented:
PR is ready for another look. I deviated from your suggested property key a bit for more flexibility and simpler implementation. Here's how it works now:
1. Base key is `kotlin.js.browser.karma.browsers` and it means "add these browsers for all js targets if they have no specific override". That way it works for both, K/JS and K/MPP plugin seamlessly.
2. Optionally, the above can be overriden for a given target via `kotlin.js.browser.karma.browsers.<targetName>` key, which then will discard the base key settings for that target.

Additionally, I've extracted away the `local.properties` logic into a new internal utility `Project::findCompositeProperty` to make it easier for potential future expansion of local properties support. It resolves property values in the following increasing priority (override) order:
1. `Project::findProperty`
2. `local.properties`
  1. `$rootDir/local.properties`
  2. `$projectDir/local.properties`
3. Command line properties (`-P`)

---

>And another mind, that in KGP we don’t have mechanism to undo particular browser.
So if in gradle.properties there is Chrome, for now it can’t be canecelled.
We don’t have something like notUseChrome(). Is it ok for your use case?

This is not an issue, since the browsers registered via these properties are additive, so turn them "off" in CI for example, one could just set the property key to an empty value via one of the overrides above (e.g. `-pkotlin.js.browser.karma.browsers=''`). Alternatively, this could only be used via local.properties files which are not to be committed to the remote.

mpetuska commented:
@ilgonmic a gentle nudge to have another look at this when you find a spare minute.

ilgonmic commented:
Hi, sorry for delay.
Basically it is ok for me, but I want to discuss usage of `local.properties`.
I am not sure that we support it properly now, and don't want to add inconsistency here.

ilgonmic commented:
Could you please consider possibility of usage `org.jetbrains.kotlin.gradle.plugin.PropertiesProvider`.
It is entry point to work with gradle properties in our gradle plugin

mpetuska commented:
Do you mean add local.properties support to your PropertiesProvider implementation or just drop local.properties all together for now?

mpetuska commented:
just so we're clear, I have no objections to dropping local.properties support within this PR as it's not the main focus here (although I still thing it would be useful). Just curious to see what you guys thing would work best at this point in timem

ilgonmic commented:
I think better to remove it support now and use Properties Provider for consistent working with Gradle properties. And in future,`local.properties` support can be just added to `PropertiesProvider`

whyoleg commented on libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/targets/js/testing/karma/KotlinKarma.kt:
```diff
@@ -93,11 +95,41 @@ class KotlinKarma(
+                "ie" -> useIe()
+                "opera" -> useOpera()
+                "phantom-js" -> usePhantomJS()
+                "safari" -> useSafari()
```
May it will be good to add else clause with an error or warning here, so misspelled words will be shown to user (with suggestion of available values), as not everyone would know all names

mpetuska replied:
Good tip, updated

ilgonmic commented:
@mpetuska `PropertiesProvider` supports `local.properties` now, but it can't override `gradle.properties` value, but can add new one. So I think that `PropertiesProvider` can be used freely

mpetuska commented:
Updated to use PropertiesProvider and removed previous local.properties implementation. Note that I've added the override behaviour still https://github.com/JetBrains/kotlin/pull/4881/files#diff-bd0ad0b8d735852a9204627fd3704dfa06dde60f3479cd58a7ca19406c4ba797R467 Let me know if you'd like me to remove that and just use regular properties behaviour.

ilgonmic commented:
Thank you!
I'd prefer to use default strategy of `PropertiesProvider` to be consistent with other properties.
Do you think it is better to align behavior?

mpetuska commented:
I guess I lust like overriding behaviour too much hah. No worries though, I'll switch it to regular props for now and you guys can reconsider overrides further down the line, should be an easy change if you choose to support it later.

ilgonmic commented:
Thank you!
I run CI build, and when it is green, I will merge it.

ilgonmic commented:
Thanks!
Merged it
